service/worker/batcher/activities.go

```diff
@@ -0,0 +1,276 @@
+	for i := 0; i < a.getOperationConcurrency(batchParams.Concurrency); i++ {
+		go startTaskProcessor(ctx, batchParams, taskCh, respCh, rateLimiter, sdkClient, a.MetricsClient, logger)
+	}
+
+	for {
+		resp, err := sdkClient.ListWorkflow(ctx, &workflowservice.ListWorkflowExecutionsRequest{
+			PageSize:      int32(pageSize),
+			NextPageToken: hbd.PageToken,
+			Query:         batchParams.Query,
+		})
+		if err != nil {
+			return HeartBeatDetails{}, err
+		}
+		batchCount := len(resp.Executions)
+		if batchCount <= 0 {
+			break
+		}
+
+		// send all tasks
+		for _, wf := range resp.Executions {
+			taskCh <- taskDetail{
+				execution: *wf.Execution,
+				attempts:  1,
+				hbd:       hbd,
+			}
+		}
+
+		succCount := 0
+		errCount := 0
+		// wait for counters indicate this batch is done
+	Loop:
+		for {
+			select {
+			case err := <-respCh:
+				if err == nil {
+					succCount++
+				} else {
+					errCount++
+				}
+				if succCount+errCount == batchCount {
+					break Loop
+				}
+			case <-ctx.Done():
+				return HeartBeatDetails{}, ctx.Err()
+			}
+		}
+
+		hbd.CurrentPage++
+		hbd.PageToken = resp.NextPageToken
+		hbd.SuccessCount += succCount
+		hbd.ErrorCount += errCount
+		activity.RecordHeartbeat(ctx, hbd)
+
+		if len(hbd.PageToken) == 0 {
+			break
+		}
+	}
+
+	return hbd, nil
+}
+
+func (a *activities) getActivityLogger(ctx context.Context) log.Logger {
+	wfInfo := activity.GetInfo(ctx)
+	return log.With(
+		a.Logger,
+		tag.WorkflowID(wfInfo.WorkflowExecution.ID),
+		tag.WorkflowRunID(wfInfo.WorkflowExecution.RunID),
+		tag.WorkflowNamespace(wfInfo.WorkflowNamespace),
+	)
+}
+
+func (a *activities) getOperationRPS(rps int) int {
+	if rps <= 0 {
+		return a.rps(a.namespace.String())
+	}
+	return rps
+}
+
+func (a *activities) getOperationConcurrency(concurrency int) int {
+	if concurrency <= 0 {
+		return a.concurrency(a.namespace.String())
+	}
+	return concurrency
+}
+
+func startTaskProcessor(
+	ctx context.Context,
+	batchParams BatchParams,
+	taskCh chan taskDetail,
+	respCh chan error,
+	limiter *rate.Limiter,
+	sdkClient sdkclient.Client,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case task := <-taskCh:
+			if isDone(ctx) {
+				return
+			}
+			var err error
+
+			switch batchParams.BatchType {
+			case BatchTypeTerminate:
+				err = processTask(ctx, limiter, task,
+					func(workflowID, runID string) error {
+						return sdkClient.TerminateWorkflow(ctx, workflowID, runID, batchParams.Reason)
+					})
+			case BatchTypeCancel:
+				err = processTask(ctx, limiter, task,
+					func(workflowID, runID string) error {
+						return sdkClient.CancelWorkflow(ctx, workflowID, runID)
+					})
+			case BatchTypeSignal:
+				err = processTask(ctx, limiter, task,
+					func(workflowID, runID string) error {
+						return sdkClient.SignalWorkflow(ctx, workflowID, runID, batchParams.SignalParams.SignalName, batchParams.SignalParams.Input)
+					})
```

Comment:
could we factor out the `processTask` call and have each case just set the `procFn` to call?

also, each func can just refer to values from `task` directly.. why do we make processTask pull values out of task and put them into these funcs?

Reply:
Just want to focus on enable per ns worker for batcher in this change. This is the current logic and I don't want to make changes to it as there is no test coverage. I can update batcher workflow logic in a different PR.